src/modules/content-s3/content-s3.c

```diff
@@ -11,22 +11,196 @@
 #if HAVE_CONFIG_H
 #include "config.h"
 #endif
+#include <jansson.h>
 #include <flux/core.h>
 
 #include "src/common/libutil/blobref.h"
 #include "src/common/libutil/log.h"
 
 #include "src/common/libcontent/content-util.h"
 
+#include "src/common/libtomlc99/toml.h"
+#include "src/common/libutil/tomltk.h"
+
+#include "src/common/libyuarel/yuarel.h"
+
 #include "s3.h"
 
 struct content_s3 {
     flux_msg_handler_t **handlers;
     struct s3_config *cfg;
     flux_t *h;
     const char *hashfun;
+    int is_secure;
+    int is_virtual_host;
 };
 
+/* Destroy module context.
+ */
+static void content_s3_destroy (struct content_s3 *ctx)
+{
+    if(ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        free (ctx->cfg);
+        free (ctx);
+        errno = saved_errno;
+    }
+
+    s3_cleanup();
+}
+
+static int parse_config (struct content_s3 *ctx,
+                         const flux_conf_t *conf,
+                         char *errbuff,
+                         int eb_size)
+{
+    flux_conf_error_t error;
+    const char *uri = NULL;
+    const char *bucket = NULL;
+    const char *cred_file = NULL;
+    char tmp[6];
+    struct yuarel yuri;
+    char *cpy;
+    struct tomltk_error toml_error;
+    toml_table_t *tbl;
+    const char *raw;
+    char *access_key;
+    char *secret_key;
+
+    if (flux_conf_unpack (conf,
+                          &error,
+                          "{s?:{s:s, s:s, s:s, s?:b} !}",
+                          "content-s3",
+                          "credential-file",
+                          &cred_file,
+                          "bucket",
+                          &bucket,
+                          "uri",
+                          &uri,
+                          "virtual-host-style",
+                          &ctx->is_virtual_host) < 0) {
+        snprintf(errbuff, eb_size, "failed to parse config file");
+        goto error;
+    }
+
+    if (!(cpy = strdup (uri)))
+        goto error;
+
+    if (yuarel_parse (&yuri, (char *) cpy) < 0) {
+        snprintf(errbuff, eb_size, "failed to parse uri");
+        goto error;
+    }
+
+    ctx->cfg->hostname = yuri.host;
+    if (!(strncat (ctx->cfg->hostname, ":", sizeof(ctx->cfg->hostname) + 1))) {
+        snprintf(errbuff, eb_size, "failed to form hostname");
+        goto error;
+    }
+
+    sprintf(tmp, "%d", yuri.port);
+    if (!(strncat (ctx->cfg->hostname, tmp, sizeof(ctx->cfg->hostname) + sizeof(yuri.port)))) {
+        snprintf(errbuff, eb_size, "failed to form hostname");
+        goto error;
+    }
+
+    if (!(cpy = strdup (bucket)))
+        goto error;
+
+    if (!(ctx->cfg->bucket = strdup (bucket)))
+        goto error;
+
+    if (!strncmp (yuri.scheme, "https", 5))
+        ctx->is_secure = 1;
+
+    if (!(tbl = tomltk_parse_file (cred_file, &toml_error))) {
+        snprintf(errbuff, eb_size, "failed to parse credentials file");
+        goto error;
+    }
```

Comment:
I'd suggest moving the cred file parsing to a separate function.

Also if the parse fails, get the parse error messages that reference the line number etc from the toml_error and put into errbuf.